### ISSUE TITLE
Cannot find an overload for "FullGroupJoin" and the argument count: "6" -- error occurs when joining an array and arraylist

### ISSUE THREAD

**Issue by Tyler-H** [ili101/Join-Object#12]
This seems similar to the issue reported and closed-norepro [here](https://github.com/ili101/Join-Object/issues/8). I recently edited my code that polls AD for all AD Groups and all the members, plus some extra member info (title and manager), because it was taking 7 hours to run. I replaced an array with an arraylist so I could replace a `+=` with `.Add()`.

In so doing, I've encountered the following error:

```powershell
Join-Object : Cannot find an overload for "FullGroupJoin" and the argument count: "6".
At C:\GetADGroups&Users.ps1:51 char:1
+ Join-Object -Left $ADGroupsList -Right $ADUsersList -LeftJoinProperty ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Join-Object], MethodException
    + FullyQualifiedErrorId : MethodCountCouldNotFindBest,Join-Object
```

For the following code:

```powershell
Import-Module -Name ActiveDirectory -Force
Import-Module -Name Join-Object -Force

#Get list of AD Groups and store it. Fairly quick; takes about 1-2 minutes to generate a CSV of ~1500 rows.
$ADGroupsList = Get-ADGroup -Filter * -Properties * | Select-Object DistinguishedName,CN,GroupCategory,Description | Sort-Object CN

#Declare ADUsersList object
$ADUsersList = New-Object -TypeName "System.Collections.ArrayList"
$ADUsersList = [System.Collections.ArrayList]@()

#Declare Record variable and set parameter schema. The [ordered] keyword ensures the data is entered in this order
$Record = [ordered] @{
    "Group Name" = ""
    "Employee Name" = ""
    "Title"= ""
    "Manager" = ""
}

#Loop through each group in the groups object. This is the really, really slow part. It currently takes ~8 hours to write ~82,600 rows. If the AD Web Service limit is set to default (5000), this will fail for a few groups as they have more than 5,000 users.
foreach ($Group in $ADGroupsList) {
    #Get list of members of the current group in the loop who are of type "user" (excludes servers or other computers, for example)
    $ArrayofMembers = Get-ADGroupMember -Identity $Group.DistinguishedName | Where { $_.objectClass -eq "user" }

    #Loop through each member in the list of members from above
    foreach ($Member in $ArrayofMembers) {
        #Get detailed user info about the current user like title and manager that aren't available from Get-ADGroupMember
        $User = Get-ADUser -Identity $Member -Properties name,title,manager | Select-Object Name, Title, @{Label="Manager";Expression={(Get-ADUser (Get-ADUser $Member -Properties Manager).Manager).Name}}
        
        #Specifies what values to apply to each property of the $Record object
        $Record."Group Name" = $Group.CN
        $Record."Employee Name" = $Member.Name
        $Record."Title" = $User.Title
        $Record."Manager" = $User.Manager

        #Put all the stored information above in a 'copy' record
        $objRecord = New-Object PSObject -property $Record

        #Add that copy to the existing data in the ADUsersList object
        [void]$ADUsersList.Add($objRecord)
    }
}

#Combines data from the two objects into one. This uses the Join-Object module from https://www.powershellgallery.com/packages/Join-Object/2.0.1
Join-Object -Left $ADGroupsList -Right $ADUsersList -LeftJoinProperty "CN" -RightJoinProperty "Group Name" -Type AllInBoth -LeftMultiMode DuplicateLines -RightMultiMode DuplicateLines -ExcludeLeftProperties DistinguishedName | Export-Csv C:\CombinedADResults1.csv -NoTypeInformation
```

This was working fine when I had $ADUsersList as a simple array (`$ADUsersList = ()@`) and then inside the inner foreach loop  had `$ADUsersList += $objRecord`, but appending with `+=` is allegedly really slow, and my script was taking over 7 hours to complete. However, it *was* completing. Once I changed the code to the above, I encountered the error at the top and came away with an empty CSV output file, instead. And it still took 7 hours :-(

Here's a sample of what the $ADGroupsList data looks like:

<pre>
DistinguishedName | CN | GroupCategory | Description
CN=1EAGroup,OU=Dept,OU=Facility,DC=nnn,DC=local | 1EAGroup | Security | Maps Departmental drive for shared documentation
</pre>

and $ADUsersList (anonymized) data:

<pre>
Group Name | Employee Name | Title | Manager
1EAGroup | Steve Rogers | Captain America | Nick Fury
</pre>

So before the change to an ArrayList object, I was getting this final output:

<pre>
CN | GroupCategory | Description | Employee Name | Title | Manager
1EAGroup | Security | Maps Departmental drive for shared documentation | Steve Rogers | Captain America | Nick Fury
</pre>

**Comment by ili101** [comment]
I didn't take into account ArrayLists only Arrays and DataTables.
I will see what is needed to add support for ArrayLists.

**Comment by Tyler-H** [comment]
Ah, that makes sense (and could be the cause of the closed issue I linked to). Thank you!

**Comment by ili101** [comment]
Can you try this new version and see if it works for you?
https://github.com/ili101/Join-Object/tree/ArrayList
You can install it automatically to your user module folder by running:
```
$Uri = 'https://raw.githubusercontent.com/ili101/Join-Object/ArrayList/Install.ps1'; & ([Scriptblock]::Create((irm $Uri))) -FromGitHub $Uri
```

**Comment by Tyler-H** [comment]
Ran into this issue trying to run that line in PowerShell

```
Failed installing module "Join-Object". Error: "A parameter cannot be found that matches parameter name 'f'." in Line 135
At line:167 char:5
+     throw ('Failed installing module "{0}". Error: "{1}" in Line {2}' ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Failed installi...'." in Line 135:String) [], RuntimeException
    + FullyQualifiedErrorId : Failed installing module "Join-Object". Error: "A parameter cannot be found that matches parameter name 'f'." in Line 135
```

Looks like for some reason it's not liking the format operator? You might need to change Line 135 to:

`Write-Verbose -Message ('Created module folder: "{0}"' -f $TargetPathSub)` (added parentheses per guidelines at https://devblogs.microsoft.com/scripting/string-formatting-in-windows-powershell/)

or something similar?

**Comment by ili101** [comment]
Sorry for that, you can try again now

**Comment by Tyler-H** [comment]
I think that resolves it. At least, I'm able to combine the ArrayList with the Array. I'll run a full test run to see if any unexpected kinks were added, but my scenario involves lots of groups & users and is not optimized, so won't know for sure until tomorrow when the script finishes.

**Comment by Tyler-H** [comment]
The full script ran successfully and the output matches exactly what I expected to see when combining two arrays. Thank you for the update to the module!

**Comment by Tyler-H** [comment]
Hi @ili101 -- Not to press you on this, but I'm curious when the enhancement you made will get included into the main repo, such that users calling `Import-Module -Name Join-Object` will get the ArrayList functionality? 

(NB - I'm completely unfamiliar with the process of getting something added/updated for PSGallery)

**Comment by ili101** [comment]
Sorry I was occupied with other things. Should be couple of clicks to update than do some checking that everything is working. I will probably do it tomorrow.

**Comment by Tyler-H** [comment]
Oh, thank you! That's great. Don't rush on my account, please!

**Comment by ili101** [comment]
Published 2.0.2 :smile:
